gr-uhd/lib/rfnoc_rx_streamer_impl.cc

```diff
@@ -0,0 +1,192 @@
+        graph, num_chans, stream_args, vlen, issue_stream_cmd_on_start));
+}
+
+
+rfnoc_rx_streamer_impl::rfnoc_rx_streamer_impl(rfnoc_graph::sptr graph,
+                                               const size_t num_chans,
+                                               const ::uhd::stream_args_t& stream_args,
+                                               const size_t vlen,
+                                               const bool issue_stream_cmd_on_start)
+    : gr::sync_block(
+          "rfnoc_rx_streamer",
+          gr::io_signature::make(0, 0, 0),
+          gr::io_signature::make(
+              num_chans,
+              num_chans,
+              ::uhd::convert::get_bytes_per_item(stream_args.cpu_format) * vlen)),
+      d_num_chans(num_chans),
+      d_itemsize(::uhd::convert::get_bytes_per_item(stream_args.cpu_format)),
+      d_vlen(vlen),
+      d_graph(graph),
+      d_stream_args(stream_args),
+      d_streamer(graph->create_rx_streamer(num_chans, stream_args)),
+      d_unique_id(
+          std::dynamic_pointer_cast<::uhd::rfnoc::node_t>(d_streamer)->get_unique_id()),
+      d_issue_stream_cmd_on_start(issue_stream_cmd_on_start)
+{
+    // nop
+}
+
+rfnoc_rx_streamer_impl::~rfnoc_rx_streamer_impl() {}
+
+
+/******************************************************************************
+ * GNU Radio API
+ *****************************************************************************/
+bool rfnoc_rx_streamer_impl::check_topology(int, int)
+{
+    GR_LOG_DEBUG(d_logger, "Committing graph...");
+    d_graph->commit();
+    return true;
+}
+
+bool rfnoc_rx_streamer_impl::start()
+{
+    if (d_issue_stream_cmd_on_start) {
+        // Start the streamers
+        ::uhd::stream_cmd_t stream_cmd(::uhd::stream_cmd_t::STREAM_MODE_START_CONTINUOUS);
+        if (d_start_time_set) {
+            stream_cmd.stream_now = false;
+            stream_cmd.time_spec = d_start_time;
+            d_start_time_set = false;
+        } else {
+            stream_cmd.stream_now = true;
+        }
+
+        GR_LOG_DEBUG(d_logger, "Sending start stream command...");
+        d_streamer->issue_stream_cmd(stream_cmd);
+    } else {
+        GR_LOG_DEBUG(d_logger, "Starting RX streamer without stream command...");
+    }
+    return true;
+}
+
+bool rfnoc_rx_streamer_impl::stop()
+{
+    // If we issue a stream command on start, we also issue it on stop
+    if (d_issue_stream_cmd_on_start) {
+        ::uhd::stream_cmd_t stream_cmd(::uhd::stream_cmd_t::STREAM_MODE_STOP_CONTINUOUS);
+        d_streamer->issue_stream_cmd(stream_cmd);
+    }
+    flush();
+    return true;
+}
+
+int rfnoc_rx_streamer_impl::work(int noutput_items,
+                                 gr_vector_const_void_star& input_items,
+                                 gr_vector_void_star& output_items)
+{
+    const size_t max_num_items_to_rx = noutput_items * d_vlen;
+    const size_t num_items_recvd =
+        d_streamer->recv(output_items, max_num_items_to_rx, d_metadata, d_timeout);
```

Comment:
no. I'm not letting Travis jump through hoops to ensure `gr-iio` not blocking in `work` explaining that gr-uhd does that for legacy reasons, since it was added prior to the self-call timeout and the opportunity to wake up the block executor after a timeout, with Ettus then re-introducing more of this.